nibabel/cmdline/diff.py

```diff
@@ -45,25 +45,41 @@ def get_opt_parser():
 
 
 def are_values_different(*values):
-    """Generically compares values, returns true if different"""
-    value0 = values[0]
-    values = values[1:]  # to ensure that the first value isn't compared with itself
+    """Generically compare values, return True if different
 
-    for value in values:
-        try:  # we sometimes don't want NaN values
-            if np.any(np.isnan(value0)) and np.any(np.isnan(value)):  # if they're both NaN
-                break
-            elif np.any(np.isnan(value0)) or np.any(np.isnan(value)):  # if only 1 is NaN
-                return True
+    Note that comparison is targetting reporting of comparison of the headers
+    so has following specifics:
+    - even a difference in data types is considered a difference, i.e. 1 != 1.0
+    - NaNs are considered to be the "same", although generally NaN != NaN  
+    """
+    value0 = values[0]
 
-        except TypeError:
-            pass
+    # to not recompute over again
+    if isinstance(value0, np.ndarray):
+        value0_nans = np.isnan(value0)
+        if not np.any(value0_nans):
+            value0_nans = None
 
+    for value in values[1:]:
         if type(value0) != type(value):  # if types are different, then we consider them different
             return True
         elif isinstance(value0, np.ndarray):
-            return np.any(value0 != value)
-
+            if value0.dtype != value.dtype or \
+               value0.shape != value.shape:
+                return True
+            # there might be NaNs and they need special treatment
+            if value0_nans is not None:
+                value_nans = np.isnan(value)
+                if np.any(value0_nans != value_nans):
+                    return True
+                if np.any(value0[np.logical_not(value0_nans)]
+                          != value[np.logical_not(value0_nans)]):
+                    return True
+            elif np.any(value0 != value):
+                return True
+        elif value0 is np.NaN:
```

Comment:
Numpy [prefers](https://docs.scipy.org/doc/numpy/reference/constants.html#numpy.NaN) you to use `np.nan`. I don't know if this means they plan to remove the synonyms in the future or if it's just a style thing. I'd go with their recommendation, though.

Also appears in the test file.